### PR TITLE
feat(std/node): Add support for process.on("exit") + refactor

### PR DESCRIPTION
**Pull request by Soremwar**
~This PR splits process into a public default-export module and a private destructurable one.~

Right now, loading the Node globals shim (std/node/global.ts) requires you to pass allow-read and allow-env even if you are not using the properties that need the special access.

The node globals shim will now use Proxied properties allowing the module to be imported without early evaluation.

### CODE REVIEW

File: std/node/_process.ts
Comment (guybedford) [inline std/node/_process.ts:72]: If the problem is that this getter doesn't support the named exports destructuring, wouldn't the better solution to be a Proxy as you say to rather get the benefit for both cases?

Comment (Soremwar) [reply std/node/_process.ts:72]: Proxy should be the solution, however a proxy-ed variable still seems to be triggering read/env operations

Comment (guybedford) [reply std/node/_process.ts:72]: How are you seeing that? The proxy should be based on an empty object base that only populates on the handler calls. I believe it's possible.

Comment (Soremwar) [reply std/node/_process.ts:72]: Perhaps, but I don't know how that would work on a named import :/

Comment (guybedford) [reply std/node/_process.ts:72]: It should be as simple as `env: new Proxy(...)` right?

Comment (Soremwar) [reply std/node/_process.ts:72]: I don't understand what you mean by that

Comment (guybedford) [reply std/node/_process.ts:72]: Something like the following:

```suggestion
  env: new Proxy(proxy_base, {
    set () {
      initProxy(envBase);
      return Reflect.set(...arguments);
    },
    ownKeys () {
      initProxy(envBase);
      return Reflect.ownKeys(envBase);
    },
    ... for defineProperty, deleteProperty, get, getOwnPropertyDescriptor, 
  }),
```

Comment (guybedford) [reply std/node/_process.ts:72]: Where the `initEnv` function does `Object.assign(envBase, Deno.env.toObject())` sort of thing.

**Comment by Soremwar** [comment]
This PR now also adds support for `process.on("exit")` by appending a listener to the `unload` event. Luckily, `unload` works pretty much the same as `on("exit")` since it won't schedule any additional work in the even loop either.

File: std/node/_process.ts
Comment (Soremwar) [reply std/node/_process.ts]: @guybedford After practicing a little with Proxies I think it's safe to say that perhaps the module separation might be the best choice. Even though proxying the properties of the env variables works to prevent an actual permission operation, the real problem comes when you try to read the env property as a whole. Since `envBase` is an empty object that doesn't have the actual env information until it's requested, logging it as a whole property will show just an empty object, deviating from the Node implementation.

Comment (guybedford) [reply std/node/_process.ts]: @Soremwar inspection should result in the `ownKeys` hook as above explicitly providing the full object. The output of console.log is hardly the normative comparison here. Even then it could be made to work with a special log symbol. `JSON.stringify` etc will work fine too.

**Comment by Soremwar** [comment]
This also solves an error in the Node implementation, since `process.argv` didn't contained the main module route as it is supposed to.

File: std/node/_process.ts
Comment (Soremwar) [reply std/node/_process.ts]: @guybedford I tried that with the `ownKeys` hook, but it doesn't seem to be triggered when you inspect the variable. Am I missing something?

Comment (guybedford) [reply std/node/_process.ts]: Inspection and usage are completely different things. Equality is not based on inspection, and inspection code paths can be adjusted using eg special symbols or any time of local inspection hook to adjust the inspection output.

Otherwise we'd never be able to use proxies at all, just because they log different doesn't mean we can't use them as replacements for other objects.

Comment (Soremwar) [reply std/node/_process.ts]: How would I go about implementing this custom log hook then? Sorry, I'm a little inexperienced in this area

I already have the proxies for argv and env so that would be the last discrepancy that needs to be taken care of

Comment (guybedford) [reply std/node/_process.ts]: @Soremwar the inspection code can be found at https://github.com/denoland/deno/blob/master/runtime/js/02_console.js#L400, where you can follow the proxy path as well. For example, a private `envBase[internal.logSymbol] = function () { return 'custom logger' }` where the `internal.logSymbol` is only accessible to Deno core code only is one approach. There are many ways to do it so you'd have to check the exact approach with another reviewer on that.

Comment (Soremwar) [reply std/node/_process.ts]: Yep, I found the function to custom inspect the variables, so it's all taken care off...sorta

As it turns out the inspect function will swallows any errors so I can't rely on it to test the permissions until #9095 gets merged

**Comment by Soremwar** [comment]
Ready for review

File: std/node/_process.ts
Comment (Soremwar) [reply std/node/_process.ts]: Nevermind, I ended up solving an unrelated bug, this was totally doable right now

File: std/node/process_test.ts
Comment (lucacasonato) [inline std/node/process_test.ts:15]: Why remove this test?

File: std/node/process.ts
Comment (kt3k) [inline std/node/process.ts:315]: Now we require tags for every TODO comment. ref: https://github.com/denoland/deno/pull/9135

File: std/node/process.ts
Comment (kt3k) [inline std/node/process.ts:1]:
```suggestion
// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
```

File: std/node/process_test.ts
Comment (Soremwar) [reply std/node/process_test.ts:15]: Previously this test was under the assumption the library would only expose things available through the ESM export (for some reason they are different to the items exposed through require), however since most Node packages used in Deno are transforms from CommonJS and modules used through createRequire not exposing this properties makes no sense

This test was used to check that the export from process was indeed different to the built in, but this is no longer the case

File: std/node/process.ts
Comment (Soremwar) [reply std/node/process.ts]: Solved

File: std/node/process.ts
Comment (Soremwar) [reply std/node/process.ts]: Done

File: std/node/process.ts
Comment (ry) [inline std/node/process.ts:89]: `Record<string, string>` ?

Comment (Soremwar) [reply std/node/process.ts:89]: Can you commit the change? I'm AFK right now

File: std/node/process.ts
Comment (bartlomieju) [inline std/node/process.ts:35]: @Soremwar FYI you can remove `deno-lint-ignore` if you add a comment to `// @ts-ignore` on why it's being used

**Review by bartlomieju — APPROVED** [review]
LGTM, thank you @Soremwar

**Pull request merged**
